File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
from collections import defaultdict

class Solution(object):
    def topKFrequent(self, nums, k):
        
        sortedArray = sorted(nums)
        hmap = defaultdict(list)
        
        count = 0 
        for index, num in enumerate(sortedArray):
            count +=1
            if(index == len(sortedArray) -1):
                hmap[count].append(num)
            elif sortedArray[index+1] != num :
                hmap[count].append(num)
                count = 0
        
        sortedCounts = sorted(hmap.keys(), reverse=True)
        
        res = [] 
        
        for count in sortedCounts:
            
            while(hmap[count]):
                res.append(hmap[count].pop())
                k-=1
                if k==0:
                    return res

        
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freqs;
        priority_queue<pair<int, int>>maxHeap;
        vector<int>res;

        for (int num : nums) {
            freqs[num] += 1;
        }

        for (auto& pair : freqs) {
            maxHeap.push({ pair.second, pair.first });
        }

        for (int i = 0; i < k; i++) {
            res.push_back(maxHeap.top().second);
            maxHeap.pop();
        }

        return res;
    }
};